File: Interviewbit/Heap And Maps/distinct-numbers-in-window.cpp
vector<int> Solution::dNums(vector<int> &A, int B) {
    int count=0;
    vector<int>ans;
    unordered_map<int,int>ump;
    for(int i=0;i<B;i++){
        ump[A[i]]++;
        if(ump[A[i]]==1) count++;
    }
    ans.push_back(count);

    for(int i=B;i<A.size();i++){
        ump[A[i-B]]--;
        if(ump[A[i-B]]==0) count--;
        ump[A[i]]++;
        if(ump[A[i]]==1) count++;
        ans.push_back(count);
    }
    return ans;
}

File: Interviewbit/Heap And Maps/distinct-numbers-in-window.java
public class Solution {
    public ArrayList<Integer> dNums(ArrayList<Integer> A, int B) {
        int count=0;
        ArrayList<Integer> ans=new ArrayList<>();
        HashMap<Integer,Integer> hashmap=new HashMap<>();
        for(int i=0;i<B;i++){
        	if(hashmap.containsKey(A.get(i))){
        		hashmap.put(A.get(i),hashmap.get(A.get(i))+1);
        	}else{
        		hashmap.put(A.get(i),1);
        		count++;
        	}
        }

        ans.add(count);
        for(int i=B;i<A.size();i++){
        	int num1=A.get(i-B);
        	int num2=hashmap.get(A.get(i-B))-1;
        	hashmap.put(num1,num2);
        	if(hashmap.get(num1)==0) count--;
        	hashmap.put(A.get(i),hashmap.getOrDefault(A.get(i),0)+1);
        	if(hashmap.get(A.get(i))==1)count++;
        	ans.add(count);
        }
        return ans;
    }
}
